src/commonMain/kotlin/arrow.exact/Exact.kt

```diff
@@ -1,23 +1,33 @@
 package arrow.exact
 
-import arrow.core.*
+import arrow.core.Either
 
-interface Exact<A, out B> {
 
-  fun from(value: A): Either<ExactError, B>
+interface Exact<A, out R : Refined<A>> : ExactEither<ExactError, A, R>
 
-  fun fromOrNull(value: A): B? {
-    return from(value).getOrNull()
-  }
+data class ExactError(val message: String)
+
+interface ExactEither<out E : Any, A, out R : Refined<A>> {
+
+  fun from(value: A): Either<E, R>
+
+  // TODO: This doesn't work for some weird reason :/
+  // fun Raise<E>.fromOrRaise(value: A): R = from(value).bind()
 
-  fun fromOrThrow(value: A): B {
-    return when (val result = from(value)) {
-      is Either.Left -> throw ExactException(result.value.message)
-      is Either.Right -> result.value
-    }
+  fun fromOrNull(value: A): R? = from(value).getOrNull()
+
+  fun fromOrThrow(value: A): R = when (val result = from(value)) {
+    is Either.Left -> throw ExactException(result.value)
+    is Either.Right -> result.value
   }
+
+  // TODO: What are your thoughts about this?
+  operator fun invoke(value: A): R = fromOrThrow(value)
 }
 
-open class ExactError(val message: String)
+class ExactException(error: Any) : IllegalArgumentException("ArrowExact error: $error")
+
+interface Refined<A> {
```

Comment:
I think there can be scenarios where users will want to have multiple values in constructor and of different type, for example

``` kotlin
val money = "10EUR"

data class Money(val amount: Amount, val currency: Currency) {
  companion object : Exact<String, Money, MoneyError>
}
```

Reply:
That's a good point 👍 Maybe we can remove the `Refined` because it isn't giving much value - it just constrains the users to call their wrapped think `value` and provides a not so useful `.map` extension. 

If we can't think of other useful `Refined<A>` extensions and benefits, we can drop the entire `Refined` interface and just stick with `Exact`. I don't have a strong opinion on this - maybe thinking about removing it

Reply:
I also was thinking about having `Refined` as a monad, but haven't found why it can be better than `Either`. `map` can be useful but for now I can't think of usage scenarios, it seems too domain specific

Reply:
+1 , I think we can't do better than `Either` and we shouldn't because people are already familiar with the `Either` concept. I'll just remove `Refined` - it's useless and prevents people from creating `Money` and other useful domain types.

Reply:
+1 I think we can drop `Refined` for now, we can always experiment and add it back later. From experience I know it's always better to be a bit more conservative with APIs 😅

If we start leveraging this library we will definitely hit use-cases that we'd want to add and I find it better to add them while we go than try to add them prematurely.

Reply:
Yup, totally agree :+1: Removed :dagger: